pkg/mutation/schema/schema.go

```diff
@@ -2,383 +2,179 @@ package schema
-			if len(ref) == 1 {
-				return nil
-			}
-			child := n.child
-			if child == nil {
-				child = &node{}
-			}
-			if err := child.add(ref[1:]); err != nil {
-				return wrapListErr(list, err)
-			}
-			n.child = child
-		default:
-			return fmt.Errorf("unknown node type: %v", t)
 		}
-		return nil
-	}()
-
-	if err != nil {
-		n.restore(backup)
-	}
-
-	return err
-}
-
-func (n *node) remove(ref []parser.Node) {
-	n.referenceCount--
-	// we don't add any schema nodes when the length
-	// of `ref` == 1, so we shouldn't remove any
-	// either
-	if len(ref) == 1 {
-		return
-	}
-	current := ref[0]
-	switch t := current.Type(); t {
-	case parser.ObjectNode:
-		obj := current.(*parser.Object)
-		if n.children == nil {
-			// no children means nothing to clean
-			return
-		}
-		child, ok := n.children[obj.Reference]
-		if !ok {
-			// child is missing, nothing to clean
-			return
-		}
-		// decrementing the reference count would remove the
-		// object, we can stop traversing the tree
-		if child.referenceCount <= 1 {
-			delete(n.children, obj.Reference)
-			return
-		}
-		child.remove(ref[1:])
-		return
-	case parser.ListNode:
-		if n.child == nil {
-			// no child, nothing to clean
-			return
+		// Only remove the conflict if all types now report there is no conflict
+		// at the path.
+		if !hasConflict {
+			delete(db.conflicts, conflictID)
 		}
-		// decrementing the reference count would remove the
-		// object, we can stop traversing the tree
-		if n.child.referenceCount <= 1 {
-			n.child = nil
-			return
-		}
-		n.child.remove(ref[1:])
-		return
-	default:
-		log.Error(fmt.Errorf("unknown node type"), "unknown node type", "node_type", t)
-		panic(fmt.Sprintf("unknown node type, schema db in unknown state: %s", t))
 	}
 }
 
-var _ error = &Error{}
-
-// Error holds errors processing the schema
-type Error struct {
-	nodeName   string
-	childError error
-}
-
-func (e *Error) Error() string {
-	builder := &strings.Builder{}
-	current := e
-	for {
-		builder.WriteString(current.nodeName)
-		child, ok := current.childError.(*Error)
-		if !ok {
-			break
-		}
-		current = child
-	}
-	builder.WriteString(": ")
-	builder.WriteString(current.childError.Error())
-	return strings.TrimPrefix(builder.String(), ".")
-}
-
-func wrapObjErr(obj *parser.Object, err error) *Error {
-	return &Error{
-		childError: err,
-		nodeName:   fmt.Sprintf(".%s", obj.Reference),
-	}
-}
-
-func wrapListErr(list *parser.List, err error) *Error {
-	var value string
-	if list.Glob {
-		value = "*"
-	} else {
-		value = fmt.Sprintf("\"%s\"", *list.KeyValue)
-	}
-	return &Error{
-		childError: err,
-		nodeName:   fmt.Sprintf("[\"%s\": %s]", list.KeyField, value),
-	}
+// HasConflicts returns true if the Mutator of the passed ID has been upserted
+// in DB and has conflicts with another Mutator. Returns false if the Mutator
+// does not exist.
+func (db *DB) HasConflicts(id types.ID) bool {
+	return db.conflicts[id]
```

Comment:
We should have a read lock here.

Reply:
Done.

Reply:
A read lock, specifically (with the mutex changed to an `RWMutex`), so that this method can be called in parallel.

That may be important if things like controllers (which run on their own thread) need to check the conflict status.

Reply:
Done.